src/qtpromise/qpromise_p.h

```diff
@@ -553,6 +556,76 @@ struct PromiseInspect
     }
 };
 
+template<typename T, typename U, typename Enable = void>
+struct PromiseConverter;
+
+template<typename T, typename U>
+struct PromiseConverter<
+    T,
+    U,
+    typename std::enable_if<!std::is_same<T, void>::value && !std::is_same<U, void>::value
+                                && !std::is_same<T, QVariant>::value
+                                && !std::is_same<U, QVariant>::value,
+                            void>::type>
+{
+    static std::function<U(const T&)> create()
+    {
+        return [](const T& value) {
+            return static_cast<U>(value);
+        };
+    }
+};
+
+template<typename T>
+struct PromiseConverter<T, void>
+{
+    static std::function<void(const T&)> create()
+    {
+        return [](const T&) {};
+    }
+};
+
+template<typename U>
+struct PromiseConverter<QVariant,
+                        U,
+                        typename std::enable_if<!std::is_same<U, void>::value, void>::type>
+{
+    static std::function<U(QVariant)> create()
+    {
+        // value is not const QVariant& because convert() is a non-const method
+        return [](QVariant value) {
+            // https://doc.qt.io/qt-5/qvariant.html#using-canconvert-and-convert-consecutively
+            if (value.canConvert<U>() && value.convert(qMetaTypeId<U>())) {
+                return value.value<U>();
+            } else {
+                throw QtPromise::QPromiseConversionException{};
+            }
+        };
+    }
+};
+
+template<typename T>
+struct PromiseConverter<T, QVariant>
+{
+    static std::function<QVariant(const T&)> create()
+    {
+        return [](const T& value) {
+            return QVariant::fromValue(value);
+        };
+    }
+};
+
+template<>
+struct PromiseConverter<void, QVariant>
+{
+    static std::function<QVariant()> create()
+    {
+        return []() {
+            return QVariant{};
```

Comment:
The only use case for casting void -> U (including QVariant) I can think of is a processing of differently typed promises in a construction like:

```
makePromiseList({QPromise<void>, QPromise<int>, QPromise<QVariant>})
.as<QString>()
.all([](const QString& val) {
  // output is QString{}, probably a QString::number(), and QVariant::toString()
});
```

Since we don't have this kind of promises list, I can't think of any other use case for such a conversion. I think we can safely drop it for now and reimplement rather quickly if there is a feature request.

`thenReturn` is an interesting approach but looks like a technicality that would require some precaution in C++, right?  If it's not a literal that is returned, then such a variable must be able to outlive the scope where the promise was created, so it's either a raw pointer, or some kind of a smart pointer captured in both lambdas. 

Anyway, for me `p.thenReturn()` sounds like there cannot be any more promise chaining after this. On the other hand, those who are familiar with bluebird might find it useful. Generally, I'm for increasing recall and recognition, so there might be a `thenReturn` in QtPromise as well.

Reply:
By removing the `void -> QVariant` cast I've got rid of `std::enable_if` completely

Reply:
`thenReturn()` would not be hard to implement since it would perform a copy (or move) of the returned value. It would also support chaining since it returns a promise:

```cpp
QPromise<T>::thenReturn(V value) const
{
    auto p = QtPromise::resolve(std::forward<V>(value));
    return then([=]() { return p; });
}

// void specialization:
QPromise<void>::thenReturn() const
{
    return then([=]() { });
}
```